_posts/2018-10-21-kubernetes-testing-e2e.markdown

```diff
@@ -0,0 +1,320 @@
+
+```
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: nginx-deployment
+  labels:
+    app: nginx
+spec:
+  replicas: 3
+  selector:
+    matchLabels:
+      app: nginx
+  template:
+    metadata:
+      labels:
+        app: nginx
+    spec: // The spec is more or less the pod object
+      containers: // As you may guess, this is a list of container definitions
+      - name: nginx 
+        image: nginx:1.15.4
+        ports:
+        - containerPort: 80	
+```
+
+A natural consequence of this is the expectation that creating a higher level object will result in the creation of lower level objects, and you can see a practical result of this in the Kubernetes test for certain resources. For example, you can clearly see this in the test for the deployment resource, located [here](https://github.com/Kubernetes/Kubernetes/blob/master/test/e2e/apps/deployment.go).
+
+I've applied the same strategy when building tests for my (admittedly, work-in-progress) [elasticsearch operator](https://github.com/matt-tyler/elasticsearch-operator/blob/master/e2e/suite/create.go). In it, I register a Custom Resource Definition. I can then create "clusters" which are objects representing a complete elasticsearch cluster. Underneath that definition though, are various lower level resources that (services, deployments, roles) that enable the Elasticsearch cluster. By watching for those resources that I expect the controller to create upon submitting a cluster definition to the controller, I can ensure the controller behaves as expected. This tends to map very well with Behavior Driven Development. I would typically start with statements like;
+
+``` 
+Given I have registered my Custom Resource Definition with the cluster,
+
+	When I apply a Custom Resource Specification
+
+		Then I should create a headless service for the control plane
+
+		Then I should create a master node deployment
+
+		Then I should create a data node deployment
+```
+
+A small demonstration: The bottom window shows the logging output of the controller, the top-left is the output of the test runner, and the top-right is a watch of on service objects (eg; `watch -n 1 kubectl get services`).
+
+![](/img/kubernetes/k8s-testing.gif)
+
+This is helping me to focus my development on the most important functionality required to get things working, without getting too bogged down in the details (which admittedly still happens but less so than before!).
+
+## Conclusion
+
+This should have given you a few ideas about how to interact with the Kubernetes API to create flexible test suites. There is a wealth of resources in the [Kubernetes community repositories](https://github.com/kubernetes/community/blob/master/contributors/devel/e2e-tests.md) if you are looking to learn more about the project's test infrastructure. I hope the examples here will serve as a useful starting point for verifying your own systems.
```

Comment:
Conclusion needs work - it's a bit weak. 

I think if you take the previous paragraph and make that the start of the conclusion ("This is helping me to focus my development...") and explain why this is true it would work better. Sell the sizzle. Remind them why they should care.